Fix a config path resolution bug on mac

awkure opened the pull request:
Problem: Currently we're using `dirs` crate to find a config
directory when reading dfx certificates. The problem is that
`dirs` authors decided to use

```
/Users/awkure/Library/Application Support/
```

as a config directory which is wrong in case of dfx.

Solution:
  * Rewrite `get_identity` to take path from $HOME
  * Make `LedgerBuilder` accept owner as `String`
  * Remove `dirs` dependency as it is no longer used

awkure committed:
Fix a config path resolution bug on mac

Problem: Currently we're using `dirs` crate to find a config
directory when reading dfx certificates. The problem is that
`dirs` authors decided to use

```
/Users/awkure/Library/Application Support/
```

as a config directory which is wrong in case of dfx.

Solution:
  * Rewrite `get_identity` to take path from $HOME
  * Make `LedgerBuilder` accept owner as `String`
  * Remove `dirs` dependency as it is no longer used

hagsteel commented on src/lib.rs:
```diff
@@ -24,14 +24,19 @@ const URL: &str = "http://localhost:8000";
-
-    let identity = BasicIdentity::from_pem_file(ident_path)?;
-    Ok(identity)
+pub fn get_identity<'a>(account_name: impl Into<&'a str>) -> Result<BasicIdentity> {
+    let home_folder = std::env::var("HOME")?;
+    let account_name: &str = account_name.into();
+    let path_string = format!("{home_folder}/.config/dfx/identity/{account_name}/identity.pem");
+    let path = Path::new(&path_string);
+    let identity = BasicIdentity::from_pem_file(path);
+    match identity {
+        Ok(identity) => Ok(identity),
+        Err(PemError::Io(e)) if e.kind() == std::io::ErrorKind::NotFound => {
+            Err(Error::CertNotFound(path.into()))
+        }
+        Err(err) => Err(Error::from(err)),
+    }
 }
```
Keep it as `Path` rather than make it into strings, as that's what `Path` is for.

Rather than removing `dirs` let's use `dirs.home_dir`.
Including the path in the error is a good idea.

```suggestion
pub fn get_identity(account_name: impl AsRef<Path>) -> Result<BasicIdentity> {
    let mut ident_path = dirs::home_dir().ok_or(crate::Error::MissingConfig)?;
    ident_path.push(".config");
    ident_path.push("dfx/identity");
    ident_path.push(account_name);
    ident_path.push("identity.pem");

    match BasicIdentity::from_pem_file(&ident_path) {
        Ok(identity) => Ok(identity),
        Err(PemError::Io(e)) if e.kind() == std::io::ErrorKind::NotFound => {
            Err(Error::CertNotFound(ident_path))
        }
        Err(err) => Err(Error::from(err)),
    }
}
```

awkure replied:
Ah, so there's a home var in dirs, ok

hagsteel commented on src/lib.rs:
````diff
@@ -44,7 +49,7 @@ fn get_identity(account_name: impl AsRef<Path>) -> Result<BasicIdentity> {
 /// cp -Rn ./identity/.config/dfx/identity/* ~/.config/dfx/identity/
 /// ```
-pub async fn get_agent(name: impl AsRef<Path>, url: Option<&str>) -> Result<Agent> {
+pub async fn get_agent(name: impl Into<&str>, url: Option<&str>) -> Result<Agent> {
````
use `Path` so it works with the `get_identity` suggestion
```suggestion
pub async fn get_agent(name: impl AsRef<Path>, url: Option<&str>) -> Result<Agent> {
```

awkure replied:
https://infinityswap.slack.com/archives/C02P69J279T/p1648709894710999?thread_ts=1648705979.062629&cid=C02P69J279T

hagsteel replied:
:+1:

hagsteel reviewed (changes requested):
We should still use `Path` for dealing with paths, I made some suggestions that should make this work for Mac as well

awkure committed:
Fix a config path resolution bug on mac

Problem: Currently we're using `dirs` crate to find a config
directory when reading dfx certificates. The problem is that
`dirs` authors decided to use

```
/Users/awkure/Library/Application Support/
```

as a config directory which is wrong in case of dfx.

Solution:
  * Rewrite `get_identity` to take path from $HOME env variable
